src/instantsearch/instantsearch.ts

```diff
@@ -178,6 +178,13 @@ export type InstantSearchConfig = {
   routing?:
     | boolean
     | {
+        router?: {
+          onUpdate?: (cb: (object) => void) => void;
+          read?: () => object;
+          write?: (routeState: object) => void;
+          createURL?: (routeState: object) => string;
+          dispose?: () => void;
```

Comment:
My bad but the attributes are required, inside InstantSearch we don't assert them before calling them. `router` & `stateMapping` [are optionals](https://github.com/algolia/instantsearch.js/blob/a35cb0b7bc0cfbd83fa98648db31f0a6d002dd2f/src/lib/InstantSearch.js#L154-L159) but not their attributes.

Reply:
👍 
I write a test as well, setting this PR as WIP

Reply:
I haven't find a way to test theses interfaces other than actually implementing an example with custom routing. https://github.com/algolia/angular-instantsearch/pull/395